[DataPipe] Moving DataPipe buffers from __iter__ to instance (self)

NivekT opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #70216
* #70373
* __->__ #76999
* #75995
* #70479
* #76776

This PR impacts `Shuffler` and `Grouper` as they previously have their buffer with the iterator rather than the instance.

Having the buffer to be tied to the object rather than the iterator will make saving the buffer during snapshotting possible.

Differential Revision: [D36216747](https://our.internmc.facebook.com/intern/diff/D36216747)

NivekT committed:
[DataPipe] Moving DataPipe buffers from __iter__ to instance (self)

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36216747).

NivekT committed:
Update on "[DataPipe] Moving DataPipe buffers from __iter__ to instance (self)"


This PR impacts `Shuffler` and `Grouper` as they previously have their buffer with the iterator rather than the instance.

Differential Revision: [D36216747](https://our.internmc.facebook.com/intern/diff/D36216747)

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36216747).

ejguan commented on torch/utils/data/datapipes/iter/combinatorics.py:
```diff
@@ -113,21 +115,46 @@ def set_shuffle(self, shuffle=True):
+
+    def __del__(self):
+        if hasattr(self, "buffer"):
+            self.buffer.clear()
```
I think `self.buffer` always exists.

NivekT replied:
Turns out it doesn't

ejguan replied:
TBH, I think this function is redundant. The `buffer` will be clean up automatically.

NivekT replied:
I will have another look. I might just remove it. I just worry about the case that `torchvision` ran into.

ejguan replied:
If that's the case, let's move `self.buffer` before assertion.

ejguan commented on torch/utils/data/datapipes/iter/grouping.py:
```diff
@@ -231,46 +233,78 @@ def __init__(self,
+            self.drop_remaining,
+            self.wrapper_class,
+        ) = state
+        self.reset()
```
Ditto

ejguan reviewed (approved):
LGTM with a few comments. We need to add `reset` to the API requirement for `DataPipe`.

NivekT commented on torch/utils/data/datapipes/iter/combinatorics.py:
```diff
@@ -113,21 +115,46 @@ def set_shuffle(self, shuffle=True):
+            self.buffer_size,
+            self._enabled
+        ) = state
+        self.reset()
```
Currently, the logic of `__getstate__` and `__setstate__`  isn't saving the content of the buffer. I am planning to add `save_snapshot` to do that, but open to do that within `__getstate__` instead.

Therefore, the reset acts as "re-initialization" here, and basically just sets the buffer back to empty.

ejguan replied:
`buffer` is never created during "re-initialization". Can I know the reason behind splitting the functionality into `__getstate__` and `save_snapshot`?

ejguan replied:
> In `DataLoader`, IIUC serialization typically is used by multiprocess and that happens in the beginning, such that the buffers are.

This is true. If we assume the buffer is empty, we don't necessarily need to run `reset` before serialization. Then, these two APIs are going to return the same thing.
If the `buffer` is not empty during serialization because of MP, I believe it should be users' intention. We should allow them to do so.

See the example from TorchVision:
https://github.com/pytorch/vision/blob/c67a58396d894c48897b96fce4eae0bcb3fd2755/torchvision/prototype/datasets/_builtin/cub200.py#L183-L185
They use `demux` to split image and meta data.
https://github.com/pytorch/vision/blob/c67a58396d894c48897b96fce4eae0bcb3fd2755/torchvision/prototype/datasets/_builtin/cub200.py#L188-L190
Then, they use the `dict` to iterate over the meta data pipeline to create a dictionary of image_id to path. At this point, the buffer inside `demux` should be full of data. If they put this data pipeline into MP, you will clean up buffer for them. This is non-intuitive to me.

NivekT replied:
Thanks for the helpful TorchVision example. The issue discussed here is pre-existing as you can see from `master` branch's demux code snippet below. In the past, we have avoided serializing the buffer (`dp2 = deserialize(serialize(dp1))` returns a `dp2` that has an empty buffer). This is true for all DataPipes with buffers, but we should definitely change it for the better if it doesn't cause other issues.
https://github.com/pytorch/pytorch/blob/9cb4e1cbf9e3506dabdb396368f2156ce3fbe6a2/torch/utils/data/datapipes/iter/combining.py#L353-L378

I think the best course of action is to add a separate PR in the stack that change DataPipes from not serializing the buffers to serializing them (often removing custom `__getstate__` and `__setstate__` should do the trick), and test those changes properly.

I think with the introduction with the "serialization wrapper DataPipe" from the other PR, it should work without issue. We can also provide guidance to users to create a custom "serialization wrapper" for DataPipes if the existing one doesn't mean their needs.

cc: @VitalyFedyunin

NivekT replied:
Actually I realize saving the buffer for `fork` and `demux` is a bad idea unless the snapshotting is complete, because without snapshotting, after the serialization/deserialization process, `main_datapipe` will read from the beginning again.

Let's say:
```python
main_dp = IterableWrapper(range(10))
cdp1, cdp2 = main_dp.fork(num_instances=2)
# cdp1, cdp2 = main_dp.demux(num_instances=2, classifier_fn=class_fn)
it1, it2 = iter(cdp1), iter(cdp2)
print(next(it1))
# Now we serialize and deserialize
new_cdp1 = pickle.loads(pickle.dumps(cdp1))
new_cdp2 = pickle.loads(pickle.dumps(cdp2))
# `main_dp` will restart from the beginning, because snapshotting isn't done yet, but the buffer will still have [0]
print(list(new_cdp1))  # We have some pointer logic that avoids element in the buffer if the child is the fastest
# [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
print(list(new_cdp2))
# [0, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
```

We will have to address this with the full snapshotting.

NivekT committed:
Update on "[DataPipe] Moving DataPipe buffers from __iter__ to instance (self)"


This PR impacts `Shuffler` and `Grouper` as they previously have their buffer with the iterator rather than the instance.

Having the buffer to be tied to the object rather than the iterator will make saving the buffer during snapshotting possible.

Differential Revision: [D36216747](https://our.internmc.facebook.com/intern/diff/D36216747)

[ghstack-poisoned]

NivekT commented on torch/utils/data/datapipes/_typing.py:
```diff
@@ -421,6 +421,7 @@ def _set_datapipe_valid_iterator_id(datapipe):
             datapipe._valid_iterator_id = 0
         else:
             datapipe._valid_iterator_id += 1
+        datapipe.reset()
```
@ejguan  `reset()` has been added here to make sure `reset()` is called every time when `__iter__` is called and updates the `_valid_iterator_id`

NivekT committed:
Update on "[DataPipe] Moving DataPipe buffers from __iter__ to instance (self)"


This PR impacts `Shuffler` and `Grouper` as they previously have their buffer with the iterator rather than the instance.

Having the buffer to be tied to the object rather than the iterator will make saving the buffer during snapshotting possible.

Differential Revision: [D36216747](https://our.internmc.facebook.com/intern/diff/D36216747)

[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Moving DataPipe buffers from __iter__ to instance (self)"


This PR impacts `Shuffler` and `Grouper` as they previously have their buffer with the iterator rather than the instance.

Having the buffer to be tied to the object rather than the iterator will make saving the buffer during snapshotting possible.

Differential Revision: [D36216747](https://our.internmc.facebook.com/intern/diff/D36216747)

[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Moving DataPipe buffers from __iter__ to instance (self)"


This PR impacts `Shuffler` and `Grouper` as they previously have their buffer with the iterator rather than the instance.

Having the buffer to be tied to the object rather than the iterator will make saving the buffer during snapshotting possible.

Differential Revision: [D36216747](https://our.internmc.facebook.com/intern/diff/D36216747)

[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Moving DataPipe buffers from __iter__ to instance (self)"


This PR impacts `Shuffler` and `Grouper` as they previously have their buffer with the iterator rather than the instance.

Having the buffer to be tied to the object rather than the iterator will make saving the buffer during snapshotting possible.

Differential Revision: [D36216747](https://our.internmc.facebook.com/intern/diff/D36216747)

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36216747).

NivekT commented:
@pytorchbot merge this please

NivekT commented:
@pytorchbot merge this please

malfet commented:
@pytorchbot merge this

NivekT commented:
@pytorchbot merge this -g

malfet commented:
@pytorchbot merge this please
